dump/pointermap.go

```diff
@@ -0,0 +1,152 @@
+}
+
+// GetPointers: Given a structure, it will recursively map all pointers mentioned in the tree,
+// breaking circular references and provide list of:
+// * all pointers
+// * all pointers that was referenced at least twice
+// * all pointers that point to primitive type
+func GetPointers(v reflect.Value) ([]uintptr, []uintptr, []uintptr){
+	pm := &pointerMap{
+		allpointers: []uintptr{},
+		reusedPointers: []uintptr{},
+		primitivePointers: []uintptr{},
+	}
+	pm.getAllAndReusedPointers(v)
+	pm.getPrimitivePointers(v)
+	return pm.allpointers, pm.reusedPointers, pm.primitivePointers
+}
+
+func (pm *pointerMap) getPrimitivePointers(v reflect.Value) {
+	if v.Kind() == reflect.Invalid {
+		return
+	}
+	// if v is a pointer pointing to a primitive value, register it as primitive pointer
+	if isPrimitivePointer(v) {
+		pm.addPrimitivePointer(v.Pointer())
+		return
+	}
+
+	// Now descend into any children of this value
+	switch v.Kind() {
+	case reflect.Slice, reflect.Array:
+		numEntries := v.Len()
+		for i := 0; i < numEntries; i++ {
+			pm.getPrimitivePointers(v.Index(i))
+		}
+
+	case reflect.Interface:
+		pm.getPrimitivePointers(v.Elem())
+
+	case reflect.Ptr:
+		pm.getPrimitivePointers(v.Elem())
+
+	case reflect.Map:
+		for _, key := range v.MapKeys() {
+			pm.getPrimitivePointers(v.MapIndex(key))
+		}
+
+	case reflect.Struct:
+		numFields := v.NumField()
+		for i := 0; i < numFields; i++ {
+			pm.getPrimitivePointers(v.Field(i))
+		}
+	}
+}
+
+// Recursively consider v and each of its children, updating pointer information
+func (pm *pointerMap) getAllAndReusedPointers(v reflect.Value) {
+	if v.Kind() == reflect.Invalid {
+		return
+	}
+	if isPointerValue(v) && v.Pointer() != 0 { // pointer is 0 for unexported fields
+		reused := pm.addPointer(v.Pointer())
+		if reused {
+			// No use descending inside this value, since it have been seen before and all its descendants
+			// have been considered
+			return
+		}
+	}
+
+	// Now descend into any children of this value
+	switch v.Kind() {
+	case reflect.Slice, reflect.Array:
+		numEntries := v.Len()
+		for i := 0; i < numEntries; i++ {
+			pm.getAllAndReusedPointers(v.Index(i))
+		}
+
+	case reflect.Interface:
+		pm.getAllAndReusedPointers(v.Elem())
+
+	case reflect.Ptr:
+		pm.getAllAndReusedPointers(v.Elem())
+
+	case reflect.Map:
+		for _, key := range v.MapKeys() {
+			pm.getAllAndReusedPointers(v.MapIndex(key))
+		}
+
+	case reflect.Struct:
+		numFields := v.NumField()
+		for i := 0; i < numFields; i++ {
+			pm.getAllAndReusedPointers(v.Field(i))
+		}
+	}
+}
+
+func (pm *pointerMap) addPrimitivePointer(ptr uintptr) {
+	for _, have := range pm.primitivePointers {
+		if ptr == have {
+			return
+		}
+	}
+	pm.primitivePointers = append(pm.primitivePointers, ptr)
+}
+
+// addPointer to the pointerMap, update reusedPointers. Returns true if pointer was reused
+func (pm *pointerMap) addPointer(ptr uintptr) bool {
+	// Is this already known to be reused?
+	for _, have := range pm.reusedPointers {
```

Comment:
Does this not cause a performance problem when you write out the AST?

Reply:
I think it does not matter so much, because even if it cause a performance problem, it's at pre-processing stage, not at runtime stage. Currently "go generate" command takes ~3 seconds on my laptop, which I think is acceptable. And we can optimize it if necessary.